utreexo-accumulator-bip.md

```diff
@@ -0,0 +1,632 @@
+  - `targets` are the positions of the elements being proven. Represented as a vector of uint64.
+  - `proof` are the hashes needed to hash the roots. Represented as a `[]hash`. `proof` MUST be in ascending order by the node positions.
+    The proof is considered invalid otherwise.
+
+# Specification
+
+The hash function SHA512/256[^2] is used for the hash operations in the accumulator.
+
+An Utreexo accumulator implementation MUST support these 3 operations: Addition, Verification, and Deletion.
+
+## Utility Functions
+
+The following utility functions are required for performing accumulator operations:
+
+**parent_hash(left, right):** Returns the hash of the concatenation of two child hashes (`left` and `right`).
+If either child is `nil`, the result is simply the non-`nil` child (treated as if the tree has a single child at that position).
+if both children are `nil`, the result is `nil`.
+
+Implementation:
+
+```python
+def parent_hash(left: bytes, right: bytes) -> bytes:
+    if right is None and left is None: return None
+    if left is None: return right
+    if right is None: return left
+
+    return sha512_256(left + right)
+```
+
+**treerows(numleaves):** Returns the minimum number of bits required to represent `numleaves - 1`. This corresponds to the height of the largest tree in the forest. Returns `0` if `numleaves` is `0`.
+
+Implementation:
+
+```python
+def treerows(numleaves: int) -> int:
+    if numleaves == 0: return 0
+    return (numleaves - 1).bit_length()
+```
+
+**is_right_sibling(position):** Returns `true` if the given `position` corresponds to a right sibling.
+A position is on the right side if its least significant bit (LSB) is set (i.e., `position & 1 == 1`).
+And it is the right sibling of a **given node** if all bits but the LSB are identical.
+
+Implementation:
+
+```python
+def is_right_sibling(position: int) -> bool:
+    return (position & 1) == 1
+```
+
+**right_sibling(position):** Returns the position of the right sibling of the given `position`.
+If `position` is already on the right side, it returns `position` unchanged.
+Otherwise, turning on the least significant bit moves the position to the right side.
+
+Implementation:
+
+```python
+def right_sibling(position: int) -> int:
+    return position | 1
+```
+
+**sibling(position):** Returns the position of the sibling of the given `position`.
+If `position` is on the right side, it returns the left sibling by turning off the least significant bit.
+If `position` is on the left side, it returns the right sibling by turning on the least significant bit.
+
+Implementation:
+
+```python
+def sibling(position: int) -> int:
+    return position ^ 1
+```
+
+**parent(position, total_rows):** Returns the parent position of the given `position` in an accumulator with `total_rows` tree rows.
+
+Implementation:
+
+```python
+def parent(position: int, total_rows: int) -> int:
+    return (position >> 1) | (1 << total_rows)
+```
```

Comment:
I could have used a little more explanation why this returns the parent, but staring at it for a bit, it seems to me that a fully filled tree with 2<super>n</super> leaves would have 2<super>n</super>-1 inner nodes, meaning that all leaves start with a zero in the first position and all inner nodes starting with a one. 

E.g. for four leaves, the leaves are 000, 001, 010, and 011, and the inner nodes would be 100, 101, 110.

For 000 and 001, shifting to the right gives 00 and setting the top bit makes the parent 100. For 010 and 011, it works out to be 101. For 100 and 101, it works out to 110.

Gotcha, cool.